src/backend/distributed/shardsplit/shardsplit_decoder.c

```diff
@@ -62,6 +75,68 @@ _PG_output_plugin_init(OutputPluginCallbacks *cb)
 	/* actual pgoutput callback will be called with the appropriate destination shard */
 	pgoutputChangeCB = cb->change_cb;
 	cb->change_cb = split_change_cb;
+	cb->filter_by_origin_cb = replication_origin_filter_cb;
+}
+
+
+/*
+ * replication_origin_filter_cb call back function filters out publication of changes
+ * originated from any other node other than the current node. This is
+ * identified by the "origin_id" of the changes. The origin_id is set to
+ * a non-zero value in the origin node as part of WAL replication.
+ */
+static bool
+replication_origin_filter_cb(LogicalDecodingContext *ctx, RepOriginId origin_id)
+{
+	if (origin_id != InvalidRepOriginId)
+	{
+		return true;
+	}
+	return false;
+}
+
+
+/*
+ * PublishChangesIfCdcSlot checks if the current slot is a CDC slot. If so, it publishes
+ * the changes as the change for the distributed table instead of shard.
+ * If not, it returns false. It also skips the Citus metadata tables.
+ */
+static bool
+PublishChangesIfCdcSlot(LogicalDecodingContext *ctx, ReorderBufferTXN *txn,
+						Relation relation, ReorderBufferChange *change)
+{
+	char *replicationSlotName = ctx->slot->data.name.data;
+
+	/* Check if the replication slot is CITUS_CDC_SLOT*/
+	if (replicationSlotName != NULL && strcmp(replicationSlotName, CITUS_CDC_SLOT_NAME) ==
+		0)
+	{
+		/* Skip publishing changes for Citus metadata tables*/
+		ObjectAddress objectAdress = { RelationRelationId, relation->rd_id, 0 };
+		if (IsObjectAddressOwnedByCitus(&objectAdress))
+		{
+			return true;
+		}
+
+		/* Check if this change is for a shard in distributed table. */
+		if (RelationIsAKnownShard(relation->rd_id))
```

Comment:
This check is also not well-optimized, because the current implementation is only used for infrequent administrator-initiated actions, and also factors in some scenarios where it cannot use caching (therefore uses slow LookupShardRelationFromCatalog call).

I think we maybe want an implementation that uses caching.

Reply:
Changed this chek to use ShardExists function instead of RelationIsAKnownShard